packages/malloy/src/model/malloy_types.ts

```diff
@@ -884,102 +968,143 @@ export interface TurtleDef extends NamedObject, Pipeline {
 }
 
-interface SubquerySource {
-  type: 'sql';
-  method: 'subquery';
-  sqlBlock: SQLBlock;
+export interface QuerySourceDef extends SourceDefBase {
+  type: 'query_source';
+  query: Query;
 }
 
-/** where does the struct come from? */
-export type StructSource =
-  | {type: 'table'; tablePath: string}
-  | {type: 'nested'}
-  | {type: 'inline'}
-  | {type: 'query'; query: Query}
-  | {type: 'sql'; method: 'nested' | 'lastStage'}
-  | {type: 'query_result'}
-  | SubquerySource;
+export interface QueryResultDef extends SourceDefBase {
+  type: 'query_result';
+}
 
-// Inline and nested tables, cannot have a StructRelationship
-//  the relationshipo is implied
+// Describes the schema which flows between pipe elements
+export interface NestSourceDef extends SourceDefBase {
+  type: 'pipeline_struct';
+  pipeSQL: string;
+}
 
-/** struct that is intrinsic to the table */
-export interface StructDef extends NamedObject, ResultStructMetadata, Filtered {
-  type: 'struct';
-  structSource: StructSource;
-  structRelationship: StructRelationship;
-  fields: FieldDef[];
-  primaryKey?: PrimaryKeyRef;
-  // "parameters in" -- values that are usable internally in this source
-  arguments?: Record<string, Argument>;
-  // "parameters out" -- values that must be passed into this source to use it
-  parameters?: Record<string, Parameter>;
-  queryTimezone?: string;
-  dialect: string;
-  annotation?: Annotation;
-  modelAnnotation?: ModelAnnotation;
+// Used by PostGres to un-JSonify at the end of a query
+export interface FinalizeSourceDef extends SourceDefBase {
+  type: 'finalize';
+}
+
+// The gesture {...sourceStruct moreProperties}  happens everywhere, now that
+// structs aren't all identical, we need a way to make one from any of the
+// exisitng structs
+
+export function sourceBase(sd: SourceDefBase): SourceDefBase {
+  return {...sd};
 }
 
+export function isSourceDef(sd: StructDef | FieldDef): sd is SourceDef {
+  return (
+    sd.type === 'table' ||
+    sd.type === 'sql_select' ||
+    sd.type === 'query_source' ||
+    sd.type === 'query_result' ||
+    sd.type === 'finalize' ||
+    sd.type === 'pipeline_struct'
+  );
+}
+
+export type SourceDef =
+  | TableSourceDef
+  | SQLSourceDef
+  | QuerySourceDef
+  | QueryResultDef
+  | NestSourceDef;
+
+/*
+ * the expression formerly known as structRelationship.type == 'base_table'
+ * mtoy todo RENAME WITHOUT CAPS WHEN YOU UNDERSTAND IT BETTER
+ */
+export function IS_BASE_TABLE(def: FieldDef | StructDef): def is SourceDef {
+  if (isJoined(def)) {
+    return false;
+  }
+  if (isSourceDef(def)) {
+    return true;
+  }
+  return false;
+}
+
+// mtoy todo NOT SURE ABOUT THIS FUNCTION EITHER
+// i.e. does this mean "joined and scalar"
+export function IS_SCALAR_ARRAY(def: FieldDef | StructDef) {
+  return def.type === 'array' && def.elementTypeDef.type !== 'record_element';
+}
+
+export type StructDef =
+  | SourceDef
+  | RecordFieldDef
+  | JoinedArrayDef
+  | FinalizeSourceDef;
```

Comment:
Should this be `FinalizeStructDef` if the distinction between a `StructDef` and a `SourceDef` is that `StructDef` things are the extra things which are allowed in the compiler?

Reply:
When I didn't understand what that structdef was, it wasn't in the source list, then it moved and i didn't rename it

Reply:
Actually this is a source and it's appearance in this list is misleading, for a long time as I was editing the PR it wasn't but I finally found the place where it was. I fixed all the places now I think.